File: tests/journal.lib/debug_cascade.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the journal
#include <pyre/journal.h>
// support
#include <cassert>


// type aliases
// chronicler
using chronicler_t = pyre::journal::chronicler_t;
// debug
using mydebug_t = pyre::journal::debug_t;
// the trash can
using trash_t = pyre::journal::trash_t;


// verify the cascade works correctly
int main() {
    // make a channel
    mydebug_t parent("test.debug.parent");
    // its activation state is what's expected
    assert(parent.active() == false);
    // it's non-fatal
    assert(parent.fatal() == false);
    // and the device is the global default
    assert(parent.device() == chronicler_t::device());
    // turn it on
    parent.activate();
    // and set the device to a trash can
    parent.device<trash_t>();

    // make a debug that's lower in the hierarchy
    mydebug_t child("test.debug.parent.blah.blah.child");
    // make sure its state is what's expected
    assert(child.active() == parent.active());
    assert(child.fatal() == parent.fatal());
    // and that it inherited the device
    assert(child.device() == parent.device());

    // all done
    return 0;
}


// end of file

File: tests/journal.api/debug_cascade.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


def test():
    """
    Verify that channels lower in the hierarchy inherit their parent default state
    """
    # get the journal
    import journal

    # make a channel
    parent = journal.debug(name="test.index.parent")
    # verify that the state is off
    assert parent.active is False
    # and that is not fatal
    assert parent.fatal is False
    # and the device is at the default value
    assert parent.device is parent.chronicler.device

    # activate it
    parent.active = True
    # and make it fatal
    parent.fatal = True
    # and set the device to a trash can
    parent.device = journal.trash()

    # lookup a name that is lower in the hierarchy
    child = journal.debug(name="test.index.parent.blah.blah.child")
    # that it's active is the same as the parent
    assert child.active == parent.active
    assert child.fatal == parent.fatal
    # and that it inherited the device correctly
    assert child.device is parent.device

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file
